lib/fs.js

```diff
@@ -2008,31 +2008,55 @@ function ReadStream(path, options) {
   this.bytesRead = 0;
   this.closed = false;
 
-  if (this.start !== undefined) {
-    if (typeof this.start !== 'number' || Number.isNaN(this.start)) {
-      throw new ERR_INVALID_ARG_TYPE('start', 'number', this.start);
-    }
-    if (this.end === undefined) {
-      this.end = Infinity;
-    } else if (typeof this.end !== 'number' || Number.isNaN(this.end)) {
-      throw new ERR_INVALID_ARG_TYPE('end', 'number', this.end);
-    }
+  // Case 0: If either start or end is undefined, set them to defaults
+  // (0, Infinity) respectively.
+  this.start = this.start === undefined ? 0 : this.start;
+  this.end = this.end === undefined ? Infinity : this.end;
 
-    if (this.start > this.end) {
-      const errVal = `{start: ${this.start}, end: ${this.end}}`;
-      throw new ERR_OUT_OF_RANGE('start', '<= "end"', errVal);
-    }
+  // Case 1: If the type of either start or end is not number, throw
+  // ERR_INVALID_ARG_TYPE (TypeError).
+  if (typeof this.start !== 'number') {
+    throw new ERR_INVALID_ARG_TYPE('start', 'number', this.start);
+  }
+  if (typeof this.end !== 'number') {
+    throw new ERR_INVALID_ARG_TYPE('end', 'number', this.end);
+  }
 
-    this.pos = this.start;
+  // Case 2: If either start or end is NaN, throw ERR_OUT_OF_RANGE (RangeError).
+  if (Number.isNaN(this.start)) {
+    throw new ERR_OUT_OF_RANGE('start', 'not NaN', this.start);
+  }
+  if (Number.isNaN(this.end)) {
+    throw new ERR_OUT_OF_RANGE('end', 'not NaN', this.end);
   }
 
-  // Backwards compatibility: Make sure `end` is a number regardless of `start`.
-  // TODO(addaleax): Make the above typecheck not depend on `start` instead.
-  // (That is a semver-major change).
-  if (typeof this.end !== 'number')
-    this.end = Infinity;
-  else if (Number.isNaN(this.end))
-    throw new ERR_INVALID_ARG_TYPE('end', 'number', this.end);
+  // Case 3: If either start or end is negative, throw ERR_OUT_OF_RANGE
+  // (RangeError).
+  if (this.start < 0) {
+    throw new ERR_OUT_OF_RANGE('start', 'positive', this.start);
+  }
+  if (this.end < 0) {
+    throw new ERR_OUT_OF_RANGE('end', 'positive', this.end);
+  }
+
+  // Case 4: If either start or end is fractional, round them to the nearest
+  // whole number. (Not integer as negatives not allowed)
+  if (!Number.isInteger(this.start)) {
```

Comment:
Now that we are doing much stricter checks, `isSafeInteger` would be better I guess.

Reply:
@thefourtheye I think that does not apply here.

Reply:
@thefourtheye I think that does not too.